gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,781 @@
+
+        while unfinished_worker_count > 0:
+            report = progress_queue.get()  # blocks if workers too slow
+            if report is None:  # a thread reporting that it finished
+                unfinished_worker_count -= 1
+                logger.info("worker thread finished; awaiting finish of %i more threads", unfinished_worker_count)
+                continue
+            examples, trained_words, raw_words = report
+            job_tally += 1
+
+            # update progress stats
+            example_count += examples
+            trained_word_count += trained_words  # only words in vocab & sampled
+            raw_word_count += raw_words
+
+            # log progress once every report_delay seconds
+            elapsed = default_timer() - start
+            if elapsed >= next_report:
+                self._log_progress(
+                    job_queue, progress_queue, cur_epoch, example_count, total_examples,
+                    raw_word_count, total_words, trained_word_count, elapsed)
+                next_report = elapsed + report_delay
+        # all done; report the final stats
+        elapsed = default_timer() - start
+        self._log_epoch_end(
+            cur_epoch, example_count, total_examples, raw_word_count, total_words,
+            trained_word_count, elapsed)
+        self.total_train_time += elapsed
+        return trained_word_count, raw_word_count, job_tally
+
+    def _train_epoch(self, data_iterable, cur_epoch=0, total_examples=None,
+                     total_words=None, queue_factor=2, report_delay=1.0):
+        """Train one epoch."""
+        job_queue = Queue(maxsize=queue_factor * self.workers)
+        progress_queue = Queue(maxsize=(queue_factor + 1) * self.workers)
+
+        workers = [
+            threading.Thread(
+                target=self._worker_loop,
+                args=(job_queue, progress_queue,))
+            for _ in xrange(self.workers)
+        ]
+
+        workers.append(threading.Thread(
+            target=self._job_producer,
+            args=(data_iterable, job_queue),
+            kwargs={'cur_epoch': cur_epoch, 'total_examples': total_examples, 'total_words': total_words}))
+
+        for thread in workers:
+            thread.daemon = True  # make interrupting the process with ctrl+c easier
+            thread.start()
+
+        trained_word_count, raw_word_count, job_tally = self._log_epoch_progress(
+            progress_queue, job_queue, cur_epoch=cur_epoch, total_examples=total_examples, total_words=total_words,
+            report_delay=report_delay)
+
+        self._set_keyedvectors()
+
+        return trained_word_count, raw_word_count, job_tally
+
+    def train(self, data_iterable, epochs=None, total_examples=None,
+              total_words=None, queue_factor=2, report_delay=1.0, callbacks=(), **kwargs):
+        """Handle multi-worker training."""
+        self._set_train_params(**kwargs)
+        if len(callbacks):
```

Comment:
What's with the `len`? I see Jayant using it all the time as well. I'm starting to think it's some Indian cultural habit. :) Why not just `if callbacks:`?

Reply:
@janpom exactly same style in Russian institutes :)